src/extensions/resolveSource.ts

```diff
@@ -10,35 +14,85 @@ export interface RegistryEntry {
+  description: string;
+  action?: string;
 }
 
 /**
- * Gets the source for a given extension name and version from the official extensions registry
+ * Gets the sourceUrl for a given extension name and version from the official extensions registry
  *
- * @param extensionName the name, or name@version of the extension to get the ExtensionSource for
+ * @param name the name of the extension to get the ExtensionSource for
  * @returns the source corresponding to extensionName in the registry
  */
-export async function resolveSource(extensionName: string): Promise<string> {
-  const [name, version] = extensionName.split("@");
+export function resolveSourceUrl(registryEntry: RegistryEntry, version: string): string {
+  const targetVersion = getTargetVersion(registryEntry, version);
+  const sourceUrl = _.get(registryEntry, ["versions", targetVersion]);
+  if (!sourceUrl) {
+    throw new FirebaseError(
+      `Could not resolve version ${clc.bold(version)} of extension ${clc.bold(registryEntry.name)}.`
+    );
+  }
+  return sourceUrl;
+}
+
+/**
+ * Looks up and returns a entry from the official extensions registry.
+ * @param name the name of the extension to get the RegistryEntry for
+ */
+export async function resolveRegistryEntry(name: string): Promise<RegistryEntry> {
   const extensionsRegistry = await getExtensionRegistry();
-  const extension = _.get(extensionsRegistry, name);
-  if (!extension) {
+  const registryEntry = _.get(extensionsRegistry, name);
+  if (!registryEntry) {
     throw new FirebaseError(`Unable to find extension source named ${clc.bold(name)}.`);
   }
+  return registryEntry;
+}
+
+/**
+ * Resolves a version or label to a version.
+ * @param registryEntry A registry entry to get the version from.
+ * @param versionOrLabel A version or label to resolve. Defaults to 'latest'.
+ */
+export function getTargetVersion(registryEntry: RegistryEntry, versionOrLabel?: string): string {
   // The version to search for when a user passes a version x.y.z or no version
-  const seekVersion = version || "latest";
+  const seekVersion = versionOrLabel || "latest";
 
   // The version to search for when a user passes a label like 'latest'
-  const versionFromLabel = _.get(extension, ["labels", seekVersion]);
+  const versionFromLabel = _.get(registryEntry, ["labels", seekVersion]);
 
-  const source =
-    _.get(extension, ["versions", seekVersion]) || _.get(extension, ["versions", versionFromLabel]);
-  if (!source) {
-    throw new FirebaseError(
-      `Could not resolve version ${clc.bold(version)} of extension ${clc.bold(name)}.`
-    );
+  return versionFromLabel || seekVersion;
+}
+
+/**
+ * Checks for and displays updateWarnings that apply to the given start and end versions.
+ * @param registryEntry the registry entry to check for updateWarnings in
+ * @param startVersion the version that you are updating from.
+ * @param endVersion the version you are updating to
+ */
+export async function checkForUpdateWarnings(
```

Comment:
This should have a better name (and updated documentation) to indicate that this will block until the user confirms the warnings (and fail out if they do not confirm).